Port code examples to C# of classes beginning with A and B

HaSa1002 opened the pull request:
First PR in my code example port effort.
See #40613 for more details.
Since there are a lot classes (and a lot less having examples) I group the changes together, but I don't really know what is practical so ¯\\_(ツ)_/¯

I created this [code translation tool](https://github.com/HaSa1002/codetranslator) to speed up the conversion process.

I left the sort as it is, because I don't know how and if it is done in C#

HaSa1002 committed:
Add C# code examples to the docs

Only existing GDScript code examples are converted and added to the
docs.
This is the first batch include classes beginning with A and B.

Included classes:
 * AcceptDialog
 * AESContext
 * Animation
 * AnimationNodeStateMachine
 * AnimationNodeStateMachinePlayback
 * AnimationNodeStateMachineTransition
 * Array
 * ArrayMesh
 * AStar
 * AStar2D
 * Bool
 * Button

clayjohn commented on doc/classes/ArrayMesh.xml:
```diff
@@ -7,19 +7,40 @@
 		m.mesh = arr_mesh
+		[/gdscript]
+		[csharp]
+		var verticies = new Godot.Collections.Array&lt;Vector3&gt;();
```
Some problems with a character here

HaSa1002 replied:
that's on purpose, because we have invalid XML otherwise (< and >)

clayjohn replied:
Ah! So it appears correctly when building the docs?

HaSa1002 replied:
yes

HaSa1002 commented:
Fixed issue in Animation.xml

Does anybody have an idea, how to parse `$Node["some/property"]`using static analysis? I might use the `$` as strong indication and/or if there is a slash in the subscription operator. Anyhow I updated the codetranslator again and made it more reliable as well as adding some features

HaSa1002 commented on doc/classes/Button.xml:
```diff
@@ -15,6 +16,20 @@
+		{
+		    var button = new Button();
+		    button.Text = "Click me";
+		    button.Connect("pressed", this, "ButtonPressed");
```
```suggestion
		    button.Connect("pressed", this, nameof(ButtonPressed));
```
That way?

Duroxxigar replied:
Yeah

HaSa1002 replied:
I think I could catch that specific case in the translation tool, but I don't know what the current output of that line is ^^

HaSa1002 replied:
Should be fixed

HaSa1002 replied:
... and is now automatically converted by the tool

aaronfranke commented on doc/classes/AStar.xml:
```diff
@@ -16,6 +17,20 @@
+		[csharp]
+		public class MyAStar : AStar
+		{
+		    public override float _ComputeCost(int u, int v)
+		    {
+		        return abs(u - v)
```
This is not valid, `abs` does not exist globally, you have to use `Mathf` for math methods. Also, you are missing a semicolon. Also, the return type should match the argument type.

```suggestion
		    public override int _ComputeCost(int u, int v)
		    {
		        return Mathf.Abs(u - v);
```

Please check your PR for similar problems (there is at least one right below this).

HaSa1002 replied:
1. Fixed and added to translation tool
2. Fixed and added to translation tool
3. This has to be fixed in the Godot API. `_compute_cost` and `_estimate_cost` are both virtual function that require a float value.

aaronfranke commented on doc/classes/Animation.xml:
```diff
@@ -6,13 +6,24 @@
+		int trackIndex = animation.AddTrack(Animation.TYPE_VALUE);
+		animation.TrackSetPath(trackIndex, "Enemy:position:x");
+		animation.TrackInsertKey(trackIndex, 0.0, 0);
+		animation.TrackInsertKey(trackIndex, 0.5, 100);
```
This might be good as-is, but double check what type this method accepts. If it's `float`, this needs to be `0.5f`.

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -227,16 +242,35 @@
+				[csharp]
+				// As there is no "in" keyword in C#, you have to use Contains
+				var array = new Godot.Collections.Array{2, 4, 6, 8};
+				if array.Contains(2) {
+				    // is true
```
Several problems. One, `if` statements always need `()`. Two, the brace style is wrong.

```suggestion
				if (array.Contains(2))
				{
				    // is true
```

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/bool.xml:
```diff
@@ -6,24 +6,50 @@
-		[/codeblock]
+		[/gdscript]
+		[csharp]
+        private bool _canShoot = true;
```
The code blocks should start with tabs in the XML documentation. There are several more problems like this.

HaSa1002 replied:
fixed. That was the one class, I converted by hand, when I implemented the feature first-hand ^^

HaSa1002 replied:
yeah that was the one class I converted by hand to test the feature when I implemented it...

HaSa1002 replied:
fixed

bojidar-bg commented on doc/classes/AESContext.xml:
```diff
@@ -35,6 +36,44 @@
+		    public var Aes = new AESContext();
+		    public override void _Ready()
+		    {
+		        var key = "My secret key!!!"; // Key must be either 16 or 32 bytes.
+		        var data = "My secret text!!"; // Data size must be multiple of 16 bytes, apply padding if needed.
```
From https://docs.godotengine.org/en/stable/getting_started/scripting/c_sharp/c_sharp_style_guide.html:
> Consider using implicitly typing (var) for declaration of a local variable, but do so only when the type is evident from the right side of the assignment [..]

HaSa1002 replied:
Idk, honestly. I just followed our C# Styleguide there.

HaSa1002 replied:
done everywhere where types weren't evident through constructor

HaSa1002 commented:
fixed a ton of `codeblock` which should have been `codeblocks`
fixed suggestions and reviewed the code again.

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -7,18 +7,33 @@
+		[/gdscript]
+		[csharp]
+		// This feature is not available in C#
+		[/csharp]
```
It's true right now, but maybe not in the near future: #42000

HaSa1002 replied:
added your example

aaronfranke commented on doc/classes/AcceptDialog.xml:
```diff
@@ -76,6 +76,7 @@
 	<signals>
 		<signal name="cancelled">
 			<description>
+			Emitted when the dialog is closed or the button created with [method add_cancel] is pressed
```
```suggestion
			Emitted when the dialog is closed or the button created with [method add_cancel] is pressed.
```

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/Animation.xml:
```diff
@@ -5,15 +5,26 @@
+		// This creates an animation that makes the node "Enemy" move to the right by
+		// 100 pixels in 0.5 seconds.
+		var animation = new Animation();
+		int trackIndex = animation.AddTrack(Animation.TYPE_VALUE);
```
```suggestion
		int trackIndex = animation.AddTrack(Animation.TrackType.Value);
```
Enums are PascalCase and are in the format `EnumName.Value` instead of `ENUMNAME_VALUE`.

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -6,20 +6,35 @@
+		var array = new Godot.Collections.Array{"One", 2, 3, "Four"}
+		GD.Print(array[0]); // One.
+		GD.Print(array[2]); // 3.
+		GD.Print(array[-1]); // Four.
```
Currently, negative indices on a `Godot.Collections.Array` crash the game with the error `System.IndexOutOfRangeException: Index was outside the bounds of the array`. We can add support for this (by wrapping with `Mathf.PosMod` in `Array`'s array index operator), but it would come at some performance cost, which might not be ideal considering many people choose C# for performance reasons. What do you think @neikeq ?

HaSa1002 replied:
My fault. I wanted to add support for this in my tool using array.Count - 1 and oversaw that. fixed

HaSa1002 replied:
Added some sort of support into my tool.

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -226,18 +241,39 @@
+				var array = new Godot.Collections.Array{2, 4, 6, 8};
+				if (array.Contains(2))
+				{
+				    // is true
```
I think it should say `GD.Print("Contains!");` or similar. You can change this in the GDScript code example too.

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/ArrayMesh.xml:
```diff
@@ -6,21 +6,42 @@
 		vertices.push_back(Vector3(0, 0, 1))
-		# Initialize the ArrayMesh.
+
+		# Initialize the ArrayMesh
```
```suggestion
		# Initialize the ArrayMesh.
```

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/ArrayMesh.xml:
```diff
@@ -6,21 +6,42 @@
 		arrays[ArrayMesh.ARRAY_VERTEX] = vertices
-		# Create the Mesh.
+
+		# Create the Mesh
```
```suggestion
		# Create the Mesh.
```
I'm not quite sure why you removed these periods.

HaSa1002 replied:
There were periods? Nonetheless, fixed.

aaronfranke commented on doc/classes/ArrayMesh.xml:
```diff
@@ -6,21 +6,42 @@
+		var arrMesh = new ArrayMesh();
+		var arrays = new Godot.Collections.Array();
+		arrays.Resize((int)ArrayMesh.ArrayType.Max);
+		arrays[(int)ArrayMesh.ArrayType.Vertex] = verticies;
```
Typo here (+ in all instances of this keyword), `verticies` should be `vertices`.

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -6,20 +6,35 @@
-		[/codeblock]
+		[/gdscript]
+		[csharp]
+		var array = new Godot.Collections.Array{"One", 2, 3, "Four"}
```
Interesting, I didn't know this syntax was possible. Anyway, you're missing a `;`

HaSa1002 replied:
fixed

aaronfranke commented on doc/classes/Button.xml:
```diff
@@ -15,7 +16,21 @@
+		    button.Text = "Click me";
+		    button.Connect("pressed", this, nameof(ButtonPressed));
+		    AddChild(button);
+		}
+		private void ButtonPressed()
```
```suggestion
		}

		private void ButtonPressed()
```

HaSa1002 replied:
fixed

HaSa1002 replied:
This fails in the static checks.

HaSa1002 replied:
Oh wait. It is the tabs. Stupid editor

aaronfranke commented on doc/classes/bool.xml:
```diff
@@ -6,24 +6,50 @@
-		[/codeblock]
+		[/gdscript]
+		[csharp]
+		private bool _canShoot = true;
```
If this is supposed to be private, you should change the GDScript code example too so it's consistent (`var _can_shoot = true`).

HaSa1002 replied:
fixed all occurrences.

aaronfranke commented:
@HaSa1002 Since #42000 was merged, you can add array concatentation. I suggest adding these lines:

```cs
// Array concatenation is not possible with C# arrays, but is with Godot.Collections.Array.
var array1 = new Godot.Collections.Array("One", 2);
var array2 = new Godot.Collections.Array(3, "Four");
GD.Print(array1 + array2); // Prints [One, 2, 3, Four]
```

HaSa1002 commented:
@aaronfranke Added the array concatentation example and did another review pass. All code examples should compile. I removed the wrong example for Array.sort `Godot.Collections.Array` has no support for it.

HaSa1002 commented on doc/classes/AnimationNodeStateMachinePlayback.xml:
```diff
@@ -6,10 +6,16 @@
 	<description>
 		Allows control of [AnimationTree] state machines created with [AnimationNodeStateMachine]. Retrieve with [code]$AnimationTree.get("parameters/playback")[/code].
```
Maybe remove the "Retrieve with .." line, since the codeblock shows how to get it. (Or just remove the `code` tag)

HaSa1002 commented on doc/classes/AnimationNodeStateMachine.xml:
```diff
@@ -6,10 +6,16 @@
 		[b]Example:[/b]
-		[codeblock]
+		[codeblocks]
+		[gdscript]
 		var state_machine = $AnimationTree.get("parameters/playback")
 		state_machine.travel("some_state")
-		[/codeblock]
+		[/gdscript]
+		[csharp]
+		var stateMachine = GetNode&lt;AnimationTree&gt;("AnimationTree").Get("parameters/playback") as AnimationNodeStateMachinePlayback;
+		stateMachine.Travel("some_state");
+		[/csharp]
+		[/codeblocks]
```
What has this example to do with the class?

aaronfranke commented on doc/classes/AStar.xml:
```diff
@@ -142,13 +171,22 @@
+				astar.AddPoint(1, new Vector3(0, 0, 0));
+				astar.AddPoint(2, new Vector3(0, 5, 0));
+				astar.ConnectPoints(1, 2);
+				Vector3 res = astar.getClosestPositionInSegment(new Vector3(3, 3, 0)); // Returns (0, 3, 0)
```
```suggestion
				Vector3 res = astar.GetClosestPositionInSegment(new Vector3(3, 3, 0)); // Returns (0, 3, 0)
```

HaSa1002 replied:
fixed

aaronfranke reviewed (changes requested):
Looks really good, found one more problem though.

aaronfranke reviewed (approved):
This should be good now. The CI failure is unrelated to this PR.

akien-mga commented:
It would need a last rebase so that we can get CI to pass (the error is on the master commit this PR is currently based on).

HaSa1002 commented:
rebased

akien-mga commented on doc/classes/AcceptDialog.xml:
```diff
@@ -76,6 +76,7 @@
 	<signals>
 		<signal name="cancelled">
 			<description>
+			Emitted when the dialog is closed or the button created with [method add_cancel] is pressed.
```
The indentation is incorrect here, it should be intended one tab further. Might be worth melding this commit into the next one too as the second commit fixes it up and includes most of the content from this PR.

HaSa1002 replied:
done.

akien-mga commented:
Thanks a ton!

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -6,20 +6,38 @@
+		GD.Print(array[2]); // 3.
+		GD.Print(array[array.Count - 1]); // Four.
+		array[2] = "Three";
+		GD.Print(array[array.Count - 2]); // Three.
```
The comment should actually be `// 3.`

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -6,20 +6,38 @@
+		[/gdscript]
+		[csharp]
+		// Array concatenation is not possible with C# arrays, but is with Godot.Collections.Array.
+		var array1 = new Godot.Collections.Array("One", 2);
+		var array2 = new Godot.Collections.Array(3, "Four");
```
Neikeq pointed out that we can also use `{` and `}`, which may be preferred...?

HaSa1002 replied:
Yeah, I switched to that syntax in later PRs

aaronfranke commented on doc/classes/Array.xml:
```diff
@@ -228,18 +246,39 @@
+				var array = new Godot.Collections.Array{2, 4, 6, 8};
+				if (array.Contains(2))
+				{
+				    GD.Print("Containes!");
```
Typo, `Containes` -> `Contains`
